### ISSUE TITLE
New table component

### ISSUE THREAD

**Issue by javfg** [opentargets/platform#1000]
### New table component

**TL;DR:** We need to improve the Table Component to enhance its functionality and compatibility.

As @andrewhercules said, it should be better visually as well as provide a convenient API that is easy to access by the different widgets implementing tables.

The list of requirements that I have gathered so far with help from the team is:

* Custom Cell contents
* Ability to fix headers as well as columns
* Droppable columns in smaller screens
* Configurable dropdown filters for columns
* Search input to help in global filtering
* Tooltipping very long headers/cells (this is tentative, but we need a way to shorten the huge titles in "related diseases" table, for example)
* Easily support serverside pagination
* Fixed height (table should not shrink height on last page when there is only a few entries)
* Multiple header rows

Paging @andrewhercules, @mkarmona, @LucaFumis and @mirandaio. If you think I am missing something, or want to add requirements, please do so. I still don't know all widgets in detail and might miss stuff that is required.

**Comment by andrewhercules** [comment]
@javfg after taking a look at the tables in the current Angular production version of the Platform and the MVP branch of the React version, here are some other requirements that immediately come to mind:

User requirements:

- Ability for user to download data table in .CSV or .TSV or JSON format
- Ability for user to sort columns in data table

Engineer/developer requirements:

- Ability for developer to set default column for sorting when table component loads
- Ability for engineer/developer to show `access_level` data attribute (note: this is used for partner/private instances to show if data source is public or private ) 
- Ability for engineer/developer to group columns to create a multi-column header

**Comment by LucaFumis** [comment]
- Ability to set the column width (probably %) to avoid cols "moving" when paginating

**Comment by d0choa** [comment]
- In angular there is a handy sorting for p-values. p-values can be passed as strings to prevent javascript rounding low p-values (~<3e-300) and still being sorted as a numeric value. This feature is critical for GWAS data (genetic variation) table

**Comment by mkarmona** [comment]
In genetics the mantissa and exponent are sent when available


--
  Regards,
  Miguel Carmona

**Comment by d0choa** [comment]
Also in the platform since 20.02, but the problem is what to do with them once you get them into the FE. The table needs to treat them as string (otherwise are rounded to 0), but still sort them numerically.

**Comment by javfg** [comment]
Sorry, misclicked on close issue. I've updated the issue with all the requirements.

**Comment by mirandaio** [comment]
Some points to consider: 

- The current [implementation](https://github.com/opentargets/ot-ui/blob/master/lib/components/OtTableRF.js) of the table allows for custom cell contents. For example: the circles in the [Assigned Genes](https://genetics.opentargets.org/variant/1_154453788_C_T) in the Genetics portal, the links and bars in the [Related Diseases](https://genetics.opentargets.org/variant/1_154453788_C_T) table of the drug page in the Platform, etc, are custom cells that display more complex stuff than raw numbers.
- The current implementation also provides the ability to decide which columns have filters and which don't. For example: the View diagram column in the [Pathways](https://platform-app.netlify.app/target/ENSG00000091831) table does not have a dropdown for filtering while the other columns do have the dropdown.
- Regarding sorting, the current implementation gives the ability to specify a column to be sorted initially and whether it should initially be ascending or descending order. In addition, you can have custom sorting. For example: the [Assigned Genes](https://genetics.opentargets.org/variant/1_154453788_C_T) is initially sorted in descending order by the Overall V2G score and you have a custom sorting that is "sorting" circles.
- I agree that it is nice that the last page of the table would be nicer if it had the same height as the other pages of the table even it has less rows. But I think it will be hard to decide what is the height of the table since rows could have different heights and we would end up with the same problem of the table changing heights as we move through the pages. It's annoying when one is going through the pages by clicking the paging arrows and having to move the mouse because the paging controls jump up and down because of the different rows sizes. This could be fixed if we agree that we make the rows fixed height but there are some cells that require being multiple lines.

**Comment by andrewhercules** [comment]
@javfg I'm moving this to the next sprint cycle - Eurovision - so we can resume work on it then

**Comment by javfg** [comment]
**Update:** So far I've implemented the following:

* Converted the table to a functional component.
* Split the code in multiple files for each part of the table (header, body, pagination, styles, sorting, utils).
* Extracted the scaffolding/container component `<PlotContainer>` that was wrapping the table. It is only used in 1 instance of the table throughout the whole project (platform and genetics) and I think it should not be inside the table (let me know if I am wrong).
* All props have sensible defaults now to make the instantation leaner. (TODO: Document the props and defaults).
* Header tooltip improved: styles can be provided from the columns array (TODO: Document).
* Colums can be shown or hidden depending on screen size breakpoints, using an API exactly like the one the [Material-UI Hidden component](https://material-ui.com/components/hidden/) implements. (See [here](https://github.com/opentargets/platform-app/commit/a77b89ac4f6fc9aecb47269cb3383c5c901bcec5#diff-8885f1927d0e7dc0a07a7c11739005eaR25)).
* Font is automatically monospaced on numeric cells, so numbers are aligned.
* Empty rows are now (almost) properly filled with whitespace in the last page of tables which do not break lines. This requires more work.

More or less halfway through the work. I will try to document everything while I finish the remaining tasks.

**Comment by javfg** [comment]
Forgot to mention data download also works for server side paginated widgets by batch-loading the dataset before export. I will include the download buttons inside the table as an option (disabled by default).
